### DISCUSSION TITLE
End the coroutine before ending the HTTP service

### DISCUSSION THREAD

**Discussion by myuid** [go-chi/chi#629] [Q&A]
I started several coroutines before http.listenandserve to monitor files and synchronize data, need to close coroutines before stopping chi, do not know how to handle, please give me some suggestions or demo.

**Comment by VojtechVitek** [comment]
Take a look at graceful http server shutdown:
https://golang.org/pkg/net/http/#Server.Shutdown

Other than that, this is very general question and doesn't really belong to chi.

**Reply by myuid** [reply]
Thank you. I'm reference https://github.com/go-chi/chi/blob/master/_examples/graceful/main.go solves this problem.

**Comment by mfridman** [comment]
Also, take a look at the resources mentioned in https://github.com/go-chi/chi/issues/627#issuecomment-857057924

At a high level, you'll need to pass a `ctx` down to your implementation, potentially wait for work to complete/timeout _and then_ call `Shutdown` on your http server.

Do check those resources though, there are code examples for you to build on.

**Reply by pkieltyka** [reply]
hey gents, I realized the _examples/graceful example was quite outdated. I made a new one here https://github.com/go-chi/chi/blob/master/_examples/graceful/main.go which is simpler and doesn't use the go-chi/valve package and instead relies on just "context" signaling and http Shutdown method.

**Reply by pkieltyka** [reply]
cc: @myuid @VojtechVitek
